examples/arduino-uno/src/bin/uno-pin-change-interrupt.rs

```diff
@@ -0,0 +1,72 @@
+/*!
+ * Example of enabling and handling pin change interrupts
+ *
+ * In this example we can get an interrupt when pin 2 changes
+ * and use that to move a stepper motor.
+ *
+ */
+#![no_std]
+#![no_main]
+#![feature(abi_avr_interrupt)]
+
+use arduino_hal;
+use panic_halt as _;
+
+use core:: sync::atomic::{AtomicBool, Ordering};
+
+static PIN_CHANGED: AtomicBool = AtomicBool::new(false);
+
+//This function is called on change of pin 2
+#[avr_device::interrupt(atmega328p)]
+#[allow(non_snake_case)]
+fn PCINT2() {
+	avr_device::interrupt::free(|_cs| {
+        PIN_CHANGED.store(true, Ordering::SeqCst);
+    });
+}
+
+#[arduino_hal::entry]
+fn main() -> ! {
+    let dp = arduino_hal::Peripherals::take().unwrap();
+    let pins = arduino_hal::pins!(dp);
+
+    //Pins used to drive the stepper motor
+    let mut dir_pin = pins.d4.into_output();
+    let mut step_pin = pins.d5.into_output();
+    
+    //Rotary encoder attached on these pins
+    let rotary_pins = [
+        pins.d2.into_floating_input().downgrade(), //CLK
+        pins.d3.into_floating_input().downgrade() //DT
+    ];
+
+    //Needed to enable pin change interrupts
+    dp.EXINT.pcicr.write(|w| unsafe { w.bits(0b100) });
+    dp.EXINT.pcmsk2.write(|w| unsafe { w.bits(0b100) });
+
+    //From this point on an interrupt can happen
+    unsafe { avr_device::interrupt::enable() };
+
+    loop {
+        avr_device::interrupt::free(|_cs| {
```

Comment:
First of all, thanks for the inspiration! This critical section seems a bit long, especially since you're already using an AtomicBool?

Reply:
If I'm using the AtomicBool I wonder do I even need to have a critical section here? I'm coming to Rust from Python so there's a lot of new concepts and I'm a bit unsure about some of it.

Reply:
A critical section is part of a bigger synchronization paradigm, like Barriers Mutexes etc. It's not just restricted to Rust. The upside of having a section where you know that your thread/process is the only one accessing a resource is that you know the ordering of your operations and their outcome. The downside is that all other threads/processes that need something from that resource have to wait. So for the overall performance of your program you want to make your time in the critical section as short as possible; take what you need and get out asap.

The AtomicBool is thread safe, but you are doing two subsequent calls to that struct on line 52 and 53. Between those two calls anything can happen, even though you expect the value to remain the same, so those need to be enclosed. However, the remainder of that if arm can be done outside of the CS, because at that point you already know you're going to execute that.

Reply:
Thanks for the explanation. I think this explains why I've seen another example where a separate function is used to read the AtomicBool, update it, and return the result. That would avoid blocking the interrupts unnecessarily.